src/model.py

```diff
@@ -0,0 +1,318 @@
+#!/usr/bin/env python3
+
+# Copyright 2023, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions
+# are met:
+#  * Redistributions of source code must retain the above copyright
+#    notice, this list of conditions and the following disclaimer.
+#  * Redistributions in binary form must reproduce the above copyright
+#    notice, this list of conditions and the following disclaimer in the
+#    documentation and/or other materials provided with the distribution.
+#  * Neither the name of NVIDIA CORPORATION nor the names of its
+#    contributors may be used to endorse or promote products derived
+#    from this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
+# EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+# PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+# CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+# EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+# PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+# OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import importlib
+import json
+import os
+
+try:
+    import torch
+except ModuleNotFoundError as error:
+    raise RuntimeError("Missing/Incomplete PyTorch package installation") from error
+
+# triton_python_backend_utils is available in every Triton Python model. You
+# need to use this module to create inference requests and responses. It also
+# contains some utility functions for extracting information from model_config
+# and converting Triton input/output types to numpy types.
+import triton_python_backend_utils as pb_utils
+
+
+def _get_model_path(config):
+    filenames = ["model.py", "model.pt"]
```

Comment:
If I am understanding the whole flow correctly:

1. The `runtime` flag for python based implementation will look for `model.py` in model/version repo, and then backend repo (in that order)
2. The python runtime for pytorch backend supports loading pytorch NN module style models as a `model.py` file

Since the core looks in model repo first, if you tried to load a pytorch `models/resnet50/1/model.py` (nn module file) that uses the Triton python runtime pytorch implementation (`/opt/tritonserver/backends/pytorch/model.py`) -- Triton core would probably try to load this `models/resnet50/1/model.py` as the backend implementation and fail, right?

Reply:
I guess this might be the most confusing part. If a model wants to load on top of Python backend, the `backend_libdir` and `backend_libpath` will initially point to the C++ runtime of the Python backend. After that, the `backend_libdir` will be "updated" to point to the Python runtime (i.e. `/opt/tritonserver/backends/pytorch/model.py`), without modifying `backend_libpath`.

The reason why the "update" logic is always correct is the Python runtime path is always assembled from `backend_dir` and `model.py`, where `backend_dir` always points to `/opt/tritonserver/backends/pytorch`, so the eventual `backend_libdir` can only be `/opt/tritonserver/backends/pytorch/model.py` for example.

Reply:
I think we should rename the runtime from `model.py` to `backend.py` to avoid any ambiguity and allow for better search logic, but `model.py` is used since TRT-LLM, so it will be a behavioral change if we do so.

Reply:
[Limit Python based backend search to backend directory](https://github.com/triton-inference-server/core/pull/285/commits/d2c456f7ba312f87da4b3dd0740643dd2a23042c)

Reply:
Issue is addressed, with commit up to https://github.com/triton-inference-server/core/pull/285/commits/f502bfc5f8c2c50d8e170a09b776ffd2f6d0d1f6 point

Reply:
Try to avoid force pushing if possible, as the [commit referenced previously](https://github.com/triton-inference-server/pytorch_backend/pull/117/files#r1407007124) is no longer valid.

Reply:
Can you clarify the current behavior now? 

For python-based backends, it will only look for `model.py` in the backend directory.

For C++ backends, will it still look in (version_dir, model_dir, backend_dir)?

Reply:
> For python-based backends, it will only look for model.py in the backend directory.
For C++ backends, will it still look in (version_dir, model_dir, backend_dir)?

Yes.

> how does this interact with escape logic?

The `backend_libdir` will be one of the (version_dir, model_dir, backend_dir) path. The `runtime` is essentially the `backend_libname`, which `<backend_libdir>/<runtime>` is the `backend_libpath`. At the end after everything are determined, there is a one time check to ensure the `backend_libpath` is within the `backend_libdir`. If the `runtime` tries to escape from the `backend_libdir` (i.e. `runtime = "../my_backend_lib.so" -> backend_libpath = "backend_dir/../my_backend_lib.so"`), this will be caught.

> can you verify running a custom backend with shared library located in the model folder still works?

Yes. I think the [L0_lifecycle](https://github.com/triton-inference-server/server/blob/main/qa/L0_lifecycle/test.sh#L1258) covers such case, and it is part of the CI run. It copies the `libtriton_identity.so` into the model folder, see `cp libtriton_identity.so models/identity_zero_1_int32/1/.` line.

Reply:
I also ran L0_passive_instance on the test container from CI and it passed, which uses a custom C++ backend: https://github.com/triton-inference-server/server/blob/main/qa/L0_passive_instance/models/distributed_int32_int32_int32/config.pbtxt#L28